### ISSUE TITLE
Train RT-DETR with ultralytics from scratch on my custom dataset

### ISSUE THREAD

**Issue by NBTailee** [ultralytics/ultralytics#8933]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

i am searching a about train RT-DETR with ultralytics from scratch on my custom dataset.
1. how can i modify code to train RT-DETR on 9 class without pretrained weights.
2. how can i download yaml file of RT-DETR
3. can you guys so me docs or demo on training RT-DETR from scratch.

**Comment by glenn-jocher** [comment]
@NBTailee hello! Thanks for reaching out with your questions about training RT-DETR with Ultralytics on your custom dataset. Here's a concise guide to get you started:

1. **Training RT-DETR on 9 Classes Without Pretrained Weights:**
   To train RT-DETR on a custom dataset with 9 classes without using pretrained weights, you'll need to modify the dataset configuration file (YAML) to specify your custom classes. Here's an example snippet for your `custom_dataset.yaml`:

   ```yaml
   # custom_dataset.yaml
   nc: 9  # number of classes
   names: ['class1', 'class2', 'class3', 'class4', 'class5', 'class6', 'class7', 'class8', 'class9']
   ```

   Then, you can start training using the following command, ensuring you specify the path to your custom dataset configuration:

   ```bash
   yolo train data=custom_dataset.yaml model=rtdetr.yaml
   ```

2. **Downloading YAML File of RT-DETR:**
   Currently, specific YAML files for RT-DETR might not be directly available for download. You would typically define your model architecture in a YAML file. For RT-DETR, you would create a `rtdetr.yaml` file that outlines the model architecture tailored to your needs. Please refer to the Ultralytics documentation for examples of model configuration files.

3. **Documentation or Demo on Training RT-DETR from Scratch:**
   For detailed documentation and demos, I recommend checking the [Ultralytics Docs](https://docs.ultralytics.com/) and the [GitHub repository](https://github.com/ultralytics/ultralytics). While specific RT-DETR training guides might not be available, the general training process for models with Ultralytics YOLOv8 should give you a solid foundation. 

   If you're new to Ultralytics YOLO, starting with the YOLOv5 or YOLOv8 training tutorials might be beneficial, as they cover the basics that apply across different models.

Feel free to join the discussions on our [GitHub Discussions page](https://github.com/ultralytics/ultralytics/discussions) if you have more questions or need further assistance. Happy training! 🚀

**Comment by NBTailee** [comment]
@glenn-jocher can i ask you a little bit more.
1. can you give a link where to download rtdetr yaml or just a example is okay.
2. what data format does rtdetr support, does it support yolo format

**Comment by glenn-jocher** [comment]
@NBTailee hi there! 👋 Sure, happy to help out.

1. For RT-DETR, you typically would define the architectures in a YAML file specific to your model's needs. While we don't have a direct RT-DETR YAML to download, you can easily start with a standard model configuration from the Ultralytics repo and adapt from there. Check the [model examples](https://github.com/ultralytics/ultralytics/tree/main/models) for a starting point!

2. Yes, RT-DETR supports the YOLO format for datasets. Ensure your dataset is organized properly: one .txt file per image, with each line in the .txt file representing a different object, formatted as `<class> <x_center> <y_center> <width> <height>`, all normalized to the image size.

I hope this helps! Let us know if you have any more questions. Happy coding! 🚀

**Comment by Yasirkhan566** [comment]
I have dataset folder having two folders images and annotations. The images folder contains train2017 and val2017. The annotation folder contains the train.json and val.json. These json files are in coco format. My dataset has only two classes. So in json file categories section has 0 label for persons and 1 for vehicles. I want to fine tune rtdetr-x.pt on my custom dataset. Please guide me how to create .yaml file and train file? If anyone can provide me the code I will be very thankful to you.

**Comment by pderrenger** [comment]
@Yasirkhan566 to fine-tune `rtdetr-x.pt` on your custom dataset, you need to create a dataset YAML file and ensure your dataset is in the correct format. Here's a brief guide:

1. Create a `custom_dataset.yaml` file with paths to your images and annotations, and specify the class names:
   ```yaml
   path: ../datasets/custom
   train: images/train2017
   val: images/val2017
   names:
     0: person
     1: vehicle
   ```

2. Convert your COCO annotations to YOLO format if necessary. You can use the `convert_coco` function from the `ultralytics.data.converter` module.

3. Train your model using the following command:
   ```bash
   yolo train data=custom_dataset.yaml model=rtdetr-x.pt
   ```

Ensure you have the latest version of the Ultralytics package to avoid any compatibility issues. If you encounter any problems, please check if they persist with the latest updates.

**Comment by Yasirkhan566** [comment]
Thanks @pderrenger

**Comment by ngocnh22** [comment]
> [@NBTailee](https://github.com/NBTailee) hi there! 👋 Sure, happy to help out.
> 
> 1. For RT-DETR, you typically would define the architectures in a YAML file specific to your model's needs. While we don't have a direct RT-DETR YAML to download, you can easily start with a standard model configuration from the Ultralytics repo and adapt from there. Check the [model examples](https://github.com/ultralytics/ultralytics/tree/main/models) for a starting point!
> 2. Yes, RT-DETR supports the YOLO format for datasets. Ensure your dataset is organized properly: one .txt file per image, with each line in the .txt file representing a different object, formatted as `<class> <x_center> <y_center> <width> <height>`, all normalized to the image size.
> 
> I hope this helps! Let us know if you have any more questions. Happy coding! 🚀

Hello @glenn-jocher, thank for your clear unscramble. Unfortunately, I dont know what steps I did wrong?
I just start to make sure the code working well but in loss log always got the 0 except at epoch = 1. I aslo attched the image for information. Remark, custom dataset yaml working well when trainied with others version YOLO. 

Thank you for your reply. 

![Image](https://github.com/user-attachments/assets/f51fe4f4-ee9c-462b-aade-c03c0684d63d)

**Comment by glenn-jocher** [comment]
@ngocnh22 Thanks for sharing your issue with RT-DETR training. The zero loss after epoch 1 is concerning and suggests a potential training issue. Here are some troubleshooting steps:

1. Try reducing your learning rate significantly (e.g., `lr0=0.0001`) as RT-DETR can be sensitive to learning rate
2. Check your batch size - RT-DETR may need a smaller batch size than other YOLO models
3. Verify your model initialization by adding `--verbose` to see initialization details
4. Make sure your dataset annotations are properly loaded by RT-DETR (try printing dataset statistics)
5. Consider training with AMP disabled using `amp=False` as mentioned in the documentation

RT-DETR uses a different architecture than traditional YOLO models, with Vision Transformers and specific loss functions, so the training dynamics can differ significantly. If these steps don't help, could you share your complete training command and model configuration for further diagnosis?
